### PR TITLE
[Xcode 9] Update Metal for GM

### PR DESCRIPTION
**Pull request by timrisi**

### CODE REVIEW

File: src/metal.cs [hunk 1]
```diff
@@ -905,6 +920,21 @@ partial interface MTLDevice {
...
 		NSString DeviceWasRemoved { get; }
 #endif
+
+		[iOS (11,0), TV (11,0), NoMac, NoWatch]
+#if XAMCORE_4_0
+		[Abstract]
+#endif
+		[Export ("newRenderPipelineStateWithTileDescriptor:options:reflection:error:")]
+		[return: NullAllowed]
+		IMTLRenderPipelineState NewRenderPipelineState (MTLTileRenderPipelineDescriptor descriptor, MTLPipelineOption options, [NullAllowed] out MTLRenderPipelineReflection reflection, [NullAllowed] out NSError error);
```
Comment (spouliot) [inline src/metal.cs]: `CreateRenderPipelineState`, e.g.

```
		[Export ("newComputePipelineStateWithFunction:options:reflection:error:")]
		IMTLComputePipelineState CreateComputePipelineState (IMTLFunction computeFunction, MTLPipelineOption options, out MTLComputePipelineReflection reflection, out NSError error);
```

File: src/metal.cs [hunk 2]
```diff
@@ -905,6 +920,21 @@ partial interface MTLDevice {
...
+
+		[iOS (11,0), TV (11,0), NoMac, NoWatch]
+#if XAMCORE_4_0
+		[Abstract]
+#endif
+		[Async (ResultTypeName="NewRenderPipelineStateResult"), Export ("newRenderPipelineStateWithTileDescriptor:options:completionHandler:")]
+		void NewRenderPipelineState (MTLTileRenderPipelineDescriptor descriptor, MTLPipelineOption options, MTLNewRenderPipelineStateWithReflectionCompletionHandler completionHandler);
```
Comment (spouliot) [inline src/metal.cs]: same

File: src/metal.cs [hunk 3]
```diff
@@ -1317,6 +1347,27 @@ partial interface MTLRenderPipelineState {
 
 		[Abstract, Export ("device")]
 		IMTLDevice Device { get; }
+
+		[iOS (11, 0), TV (11,0), NoMac, NoWatch]
+		[Abstract]
```
Comment (spouliot) [inline src/metal.cs]: that's an existing type, you cannot add `abstract` requirement on it

File: src/metal.cs [hunk 4]
```diff
@@ -1317,6 +1347,27 @@ partial interface MTLRenderPipelineState {
...
+		[Export ("maxTotalThreadsPerThreadgroup")]
+		nuint MaxTotalThreadsPerThreadgroup { get; }
+
+		[iOS (11, 0), TV (11,0), NoMac, NoWatch]
+		[Abstract]
```
Comment (spouliot) [inline src/metal.cs]: same

File: src/metal.cs [hunk 5]
```diff
@@ -1317,6 +1347,27 @@ partial interface MTLRenderPipelineState {
...
+		[Export ("threadgroupSizeMatchesTileSize")]
+		bool ThreadgroupSizeMatchesTileSize { get; }
+
+		[iOS (11, 0), TV (11,0), NoMac, NoWatch]
+		[Abstract]
```
Comment (spouliot) [inline src/metal.cs]: same

File: src/metal.cs [hunk 6]
```diff
@@ -1317,6 +1347,27 @@ partial interface MTLRenderPipelineState {
...
+		[Export ("imageblockSampleLength")]
+		nuint ImageblockSampleLength { get; }
+
+		[iOS (11,0), TV (11,0), NoMac, NoWatch]
+		[Abstract]
```
Comment (spouliot) [inline src/metal.cs]: same

**Comment by spouliot** [comment]
@timrisi please double check the test failures

File: src/metal.cs [hunk 7]
```diff
@@ -905,6 +920,21 @@ partial interface MTLDevice {
...
+#endif
+		[Export ("newRenderPipelineStateWithTileDescriptor:options:reflection:error:")]
+		[return: NullAllowed]
+		IMTLRenderPipelineState CreateRenderPipelineState (MTLTileRenderPipelineDescriptor descriptor, MTLPipelineOption options, [NullAllowed] out MTLRenderPipelineReflection reflection, [NullAllowed] out NSError error);
+
+		[iOS (11,0), TV (11,0), NoMac, NoWatch]
+#if XAMCORE_4_0
+		[Abstract]
+#endif
+		[Async (ResultTypeName="NewRenderPipelineStateResult"), Export ("newRenderPipelineStateWithTileDescriptor:options:completionHandler:")]
```
Comment (dalexsoto) [inline src/metal.cs]: Remove `Async`, it is not currently possible to do in Protocols

**Comment by VincentDondain** [comment]
Not sure we want to block the PR for that but we really should test `MTLRenderCommandEncoder`, we're adding manual code without tests here.

File: src/metal.cs [hunk 8]
```diff
@@ -447,8 +449,7 @@ partial interface MTLComputeCommandEncoder : MTLCommandEncoder {
 		[Export ("waitForFence:")]
 		void Wait (IMTLFence fence);
 
-		[Mac (10,13, onlyOn64: true)]
-		[NoTV][NoiOS]
+		[Mac (10,13, onlyOn64: true), iOS (11,0), TV (11,0)]
```
Comment (VincentDondain) [inline src/metal.cs]: Minor but I try no to update existing availability attributes so I would have added `[iOS (11,0), TV (11,0)]` bellow the Mac one in replacement of `[NoTV][NoiOS]` (;

**Comment by spouliot** [comment]
build timed out (after 2 hours stuck)

**Comment by spouliot** [comment]
build

**Pull request merged**